src/treetensornetwork/projttno.jl

```diff
@@ -0,0 +1,52 @@
+"""
+ProjTTNO
+"""
+mutable struct ProjTTNO <: AbstractProjTTNO
```

Comment:
I think it is better to not make this `mutable`, in general it is easier to reason about non-mutable structs.

If we need to update `pos` in-place, we can wrap it in a `Ref`.

Reply:
I updated things so that all tree tensor network types are immutable. I can get away with not updating the orthogonality center (for `TTNS` and `TTNO`) or position (for `ProjTTNO`, `ProjTTNOSum` as well as relatives in `ITensorTDVP.jl`) in-place, as long as things such as truncation, orthogonalization, local updates in TDVP, ..., are also out of place. We can maybe discuss whether this is fine or whether we do want to be able to update in-place via a `Ref`.

Reply:
In general, I have found it is best to avoid in-place operations unless there is a performance reason for using it, since it makes the code easier to reason about (and also makes it easier for automatic differentiation). So I would say try to avoid in-place operations in general, and then we can profile things and see if anything is slow because of that. Because most of our code is going to be dominated by tensor contractions and decompositions, we can generally get away with some extraneous shallow copying of high level data structures (like graphs, dictionaries, etc.).